### ISSUE TITLE
Clip text in Box

### ISSUE THREAD

**Issue by szsoftware** [typst/typst#2092]
Is there any way to clip text (and probably elliptize the cut text) inside a box?

I'm currently writing a pseudo-form library for typst to design forms having checkboxes, textfields (multiline and single line).
The values can be templated reading json or other formats.
Thus I can display checked boxes or filled "inputs".

There is no effect to text when having single line, specified with an appropriate height using that code:

`#let textfield(
  text: "",
  width: 5cm,
  height: 0.7cm,
  padding: 2pt
) = {
  box({
      align(horizon, {
        box(
          width:  width,
          height: height,
          clip: true,
          stroke: black,
          inset: padding,
          text
        )
       ...
`

**Comment by Enivex** [comment]
Can you produce a minimal reproducible example of what you mean?

```#let textfield(
  text: "",
  width: 5cm,
  height: 0.7cm,
  padding: 2pt ) = box( width:  width, height: height, clip: true, stroke: black, inset: padding, text
)

#textfield(text: lorem(50))```

produces

![image](https://github.com/typst/typst/assets/2191754/2de25c45-f82b-46f0-91d9-944604790673)

Is that not what you want?

(I removed the outermost box and align)

**Comment by szsoftware** [comment]
> Can you produce a minimal reproducible example of what you mean?
> 
> ```
> #let textfield(
>   text: "",
>   width: 5cm,
>   height: 0.7cm,
>   padding: 2pt
> ) = box(
>   width:  width,
>   height: height,
>   clip: true,
>   stroke: black,
>   inset: padding,
>   text
> )
> 
> #textfield(text: lorem(50))
> ```
> 
> produces
> 
> ![image](https://user-images.githubusercontent.com/2191754/266782809-2de25c45-f82b-46f0-91d9-944604790673.png)
> 
> Is that not what you want?

No, I want to have a clear "one-lined" text. Overflowing text cut or with ellipse (...)

> 
> (I removed the outermost box and align)
> 
> Edit: As an aside, a much easier way of accomplishing the same thing is

Thanks, will use it more in future. Here I would like to encapsulate the textfield element and it's properties and hide that of a box.

> 
> ```
> #let textfield = box.with(
>   width: 5cm,
>   height: 0.7cm,
>   inset: 2pt,
>   clip: true,
>   stroke: black
> )
> ```

Here's my work of the last hour:

![typst_textfield_with_elipse](https://github.com/typst/typst/assets/2539622/ea871a16-6fdb-443e-9e73-078c7142cbd6)

Textfield.typ:

`#let textEllipse(text, maxSize, placeHolder: "") = {

  let clip(text) = {
    return text.slice(0, text.len() -1).trim()
  }

  style(styles => {
    if measure(text + placeHolder, styles).width < maxSize {
      return text + placeHolder
    } else {
      textEllipse(
        clip(text),
        maxSize,
        placeHolder: "..."
      )
    }
  })
}


#let textfield(
  caption: "",
  text: "",
  width: 1cm,
  height: 0.7cm,
  padding: 2pt
) = {
  box({
      align(horizon, {
        box(
          width:  width,
          height: height,
          stroke: black,
          inset: padding,
          textEllipse(text, width - padding)
        )
        //place(start + horizon, {
        //  //h(size + 2pt + space)
        //  caption
        //})
      })
  })
}`

Calling side:

`#import "../common/form/textfield.typ": textfield

#textfield(
  width: 3cm,
  text: "This happens when"
)`

So far only width, thus one line is supported.

**Comment by szsoftware** [comment]
Btw., how can I post typst code to github issues, properly formatted? Code tags don't seem to work as expected.

**Comment by laurmaedje** [comment]
> Btw., how can I post typst code to github issues, properly formatted? Code tags don't seem to work as expected.

You need to use three backticks at the start on a separate line and the same at the end.
